### PR TITLE
[TRTLLM-12453][fix] Accommodate chunked prefill in Nemotron's EVS merging logic

### PR DESCRIPTION
**Pull request by moraxu**
## Summary by CodeRabbit

* **Bug Fixes**
  * More accurate multimodal context handling: video placeholders expand to the correct number of image tokens and audio tokens pass through intact.
  * Safer chunked prefill behavior with runtime-aware validation and token accounting; preserves generation tail and prevents out-of-range context writes.
* **Tests**
  * Added tests covering chunked prefill, mixed audio+video batches, and runtime-aware multimodal processing.

## Description

EVS MM token merging has not been chunked-prefill aware.

In `modeling_nemotron_nano.py`, `merge_evs_mm_embeds` now:
- rebuilds the full post-EVS context per multimodal request
- uses `multimodal_runtime.past_seen_token_num:chunk_end_pos` to splice only the active chunk into `input_ids`
- preserves any generation tokens after the context chunk
- refreshes runtime MM-token counters from the post-EVS layout so `find_input_mm_embeds` slices embeddings consistently

## Test Coverage

Ran the full Video-MME accuracy benchmarks.

1. With EVS & chunked prefill:
```
            "--reasoning_parser", "nano-v3",
            "--max_batch_size", "64",
            "--max_num_tokens", "64",
            "--enable_chunked_prefill",
            "--video_pruning_rate", "0.5",
```
Results:
```
    "short": {
        "overall": "0.648",
[...]
    "medium": {
        "overall": "0.530",
[...]
    "long": {
        "overall": "0.498",
[...]
    "overall": {
        "overall": "0.559",
```

2. With EVS & without chunked prefill:
```
            "--reasoning_parser", "nano-v3",
            "--max_batch_size", "64",
            "--video_pruning_rate", "0.5",
```
Results:
```
    "short": {
        "overall": "0.652",
[...]
    "medium": {
        "overall": "0.526",
[...]
    "long": {
        "overall": "0.498",
[...]
    "overall": {
        "overall": "0.559",
```

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by moraxu** [comment]
Assigning @2ez4bz and @venkywonka for viz

**Comment by moraxu** [comment]
@coderabbitai full review

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py [hunk 1]
```diff
@@ -2514,52 +2531,104 @@ def merge_evs_mm_embeds(
...
+        if context_ids.shape[0] > input_ids.shape[0]:
+            raise ValueError(
+                "EVS-adjusted context length "
+                f"({context_ids.shape[0]}) exceeds input_ids length "
+                f"({input_ids.shape[0]}). This usually means a chunked "
+                "multimodal request reached merge_evs_mm_embeds without "
+                "multimodal_runtime metadata."
+            )
+        context_ids = context_ids.to(device=input_ids.device, dtype=input_ids.dtype)
         input_ids[: context_ids.shape[0]] = context_ids
```
Comment (venkywonka) [inline tensorrt_llm/_torch/models/modeling_nemotron_nano.py:2708]: this might be unrelated to ur changes, but what happens when we have a text-only request in the same flattened batch as our EVS request? will correspond to the first text-only request (1) - `input_ids[:req_1]` - and then the EVS request (2) would be in `input_ids[req1:req2]` right? ... so won't we be overwriting it when we do `input_ids[: context_ids.shape[0]] = context_ids`? 

for eg consider this scenario: 
```
 scheduled context_requests = [A text-only chunk, B EVS video chunk]
  input_ids                   = [A ids][B ids][generation ids]
  multimodal_params           = [B only]
  merge_evs_mm_embeds builds  = [B adjusted ids]
  write                       = input_ids[:len(B adjusted ids)] = B adjusted ids
```
that clobbers the start of A in `input_ids` right?

maybe this situation is unreachable - but wanted to check

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py [hunk 2]
```diff
@@ -1286,12 +1286,26 @@ def expand_prompt_token_ids_for_mm(
...
         if has_audio:
             expanded = self._expand_audio_placeholders_in_token_ids(
                 prompt_token_ids, num_mm_tokens_per_placeholder
             )
-            return expanded, None
+            mm_data_updates = None
+            if self.video_pruning_rate > 0:
+                mm_data_updates = {
+                    "audio": {
+                        "evs_ids": torch.tensor(expanded, dtype=torch.long),
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_nemotron_nano.py:1305]: Why does audio need updates for EVS?

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py [hunk 3]
```diff
@@ -2514,52 +2531,104 @@ def merge_evs_mm_embeds(
...
-                        fill_value=self.img_context_token_id,
-                        dtype=evs_ids.dtype,
-                        device=evs_ids.device,
+            # Image/audio modalities: keep input_ids unchanged during inflight-batching.
+            if modality in ("image", "audio"):
+                context_ids = evs_ids
+            elif modality == "video":
+                # evs_ids is a flat 1-D tensor built at token-ID level.
+                # Find placeholder positions and replace each with the EVS count.
+                request_context_parts = []
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_nemotron_nano.py]: Nit: for readability, can the part that builds the `context_ids` and the one that adjusts the `multimodal_param.multimodal_runtime` be put in their own helper methods?

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py (hunk 1 above)
Comment (moraxu) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py:2708]: Yes, that is indeed another bug.. Good catch. I ran it with Codex who suggested to have `model_engine.py` compute start/end offsets for the current chunk, because it owns the flattened `input_ids` construction. In the context loop, right before it does `input_ids.extend(prompt_tokens)`, it can record:
```
start = len(input_ids)
input_ids.extend(prompt_tokens)
end = len(input_ids)
```
If that request has multimodal content, we could stash (start, end) on its `MultimodalRuntimeData` or `multimodal_data`. Those would be flattened current offsets for this chunk only. Then `merge_evs_mm_embeds` would write each adjusted chunk into its own span:
```
input_ids[start:end] = context_ids
```
So `runtime.past_seen_token_num/chunk_end_pos` still mean prompt-local chunk positions, while start/end mean flattened-batch positions. I could handle it in a different PR or this one?

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py (hunk 2 above)
Comment (moraxu) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py:1305]: As discussed offline, it's related to https://github.com/NVIDIA/TensorRT-LLM/blob/95385b85fdbf178d3ae7ae60bd93c8cbd1d64203/tensorrt_llm/inputs/registry.py#L875

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py (hunk 3 above)
Comment (moraxu) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py]: Done

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py (hunk 2 above)
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py:1305]: Sorry for being thick, but I'm still unclear why audio would need it. Let's discuss more in person.

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py (hunk 1 above)
Comment (venkywonka) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py:2708]: since its related, better to handle it in this PR itself?

Comment (moraxu) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py:2708]: Sorry, just saw this comment - we agreed with @2ez4bz to create another jira for this, since this might be a potentially larger change..

**Review by 2ez4bz — APPROVED** [review]
